caluma/caluma_form/domain_logic.py

```diff
@@ -276,7 +301,42 @@ def create(
         document.meta.pop("_defer_calculation", None)
         document.save()
 
-        recalculate_answers_from_document(document)
+        root_doc = document.family
+        root_doc = (
+            models.Document.objects.filter(pk=document.family_id)
+            .prefetch_related(
+                *utils.build_document_prefetch_statements(prefetch_options=True)
+            )
+            .first()
+        )
+        struc = structure.FieldSet(root_doc, root_doc.form)
+
+        # Initialize all calculated questions in the form.
+        # In order to do this efficiently, we get all calculated questions with their dependents,
+        # sort them topoligically, and then update their answer.
+        calculated_questions = (
+            models.Form.get_all_questions([(document.family or document).form_id])
+            .filter(type=models.Question.TYPE_CALCULATED_FLOAT)
+            .values("slug", "calc_dependents")
+        )
+        adjacency_list = {
+            dep["slug"]: dep["calc_dependents"] for dep in calculated_questions
+        }
+        ts = TopologicalSorter(adjacency_list)
+        # TopologicalSorter expects the adjacency_list the "other way around", i.e.
+        # for every node the incoming nodes should be given. To account for this, we
+        # just reverse the resulting order.
+        sorted_question_slugs = list(reversed(list(ts.static_order())))
+
+        # fetch all related questions in one query, but iterate according
+        # to pre-established sorting
+        _questions = models.Question.objects.in_bulk(sorted_question_slugs)
+        for slug in sorted_question_slugs:
+            print("question", slug)
+            update_or_create_calc_answer(
+                _questions[slug], document, struc, update_dependents=False
+            )
+
```

Comment:
I'd like to have this factored out into it's own (private if needed) method

Reply:
done